### PR TITLE
Check for loops in recursive copy

### LINKED ISSUE #43: MoreFiles.copyRecursive loops infinitely until failure when copying a directory into one of its subdirectories
Repro:
Path a = Paths.get("a") // a is an existing directory with contents
Path b = Paths.get("a/b") // b does not exist in a
MoreFiles.copyRecursive(a, b)

This loops infinitely until failure because the copy generates more content to copy. It also passes both precondition checks in MoreFiles.copyRecursive because "a/b" does not exist AND "a" does exist.

### PR DESCRIPTION
**Pull request by bluekeyes**
Fixes #43. This is arguably an API-breaking change because it also changes the exceptions thrown under certain error conditions. However, since the signature does not change and we do not expect people to catch `IllegalArgumentException`, this should be safe.

### CODE REVIEW

File: fs-base/src/main/java/com/palantir/giraffe/file/MoreFiles.java [hunk 1]
```diff
@@ -248,15 +250,25 @@ public static void copyLarge(Path source, Path target) throws IOException {
      * @param source the path to copy
      * @param target the target path
      *
-     * @throws IllegalArgumentException if the destination already exists or one
-     *         or more parent directories do not exist.
+     * @throws FileAlreadyExistsException if the target already exists
+     * @throws NoSuchFileException if and parents of the target do not exist
```
Comment (bluekeyes) [inline fs-base/src/main/java/com/palantir/giraffe/file/MoreFiles.java]: "if any"

File: fs-base/src/main/java/com/palantir/giraffe/file/MoreFiles.java [hunk 2]
```diff
@@ -248,15 +250,25 @@ public static void copyLarge(Path source, Path target) throws IOException {
...
-                "parent directory for target (%s) does not exist.", target.toAbsolutePath());
+        if (Files.exists(target)) {
+            throw new FileAlreadyExistsException(target.toString());
+        }
+
+        if (!Files.exists(target.getParent())) {
+            throw new NoSuchFileException(target.toString());
+        }
+
+        if (target.startsWith(source)) {
```
Comment (ajlake) [inline fs-base/src/main/java/com/palantir/giraffe/file/MoreFiles.java]: Is this check susceptible to funny business?

what if I do
source = Paths.get(".")
target = source.resolve("foo").toAbsolutePath().normalize()
then
will target.startsWith(source) be true?

**Comment by bluekeyes** [comment]
@ajlake the latest update should fix your issue (confirmed locally)

**Comment by ajlake** [comment]
LGTM

**Pull request merged**